src/lang/builtins_source.ml

```diff
@@ -109,3 +109,42 @@ let () =
         Clock.force_init (fun x -> List.exists (fun y -> Oo.id x = Oo.id y) l)
       in
       Lang.list ~t:s_t (List.map (fun x -> Lang.source (x :> Source.source)) l))
+
+let () =
+  let kind = Lang.univ_t () in
+  add_builtin "source.dump" ~cat:Liq
+    ~descr:"Immediately encode the whole contents of a source into a file."
+    ~flags:[Lang.Experimental]
+    [
+      ("", Lang.format_t kind, None, Some "Encoding format.");
+      ("", Lang.string_t, None, Some "Name of the file.");
+      ("", Lang.source_t (Lang.univ_t ()), None, Some "Source to encode");
+    ]
+    Lang.unit_t
+    (fun p ->
+      let format_val = Lang.assoc "" 1 p in
+      let format = Lang.to_format format_val in
+      let encoder =
+        try Encoder.get_factory format
+        with Not_found ->
+          raise (Lang_errors.Invalid_value (format_val, "Unsupported format"))
+      in
+      let kind = Encoder.kind_of_format format in
+      let file = Lang.to_string (Lang.assoc "" 2 p) in
+      let s = Lang.to_source (Lang.assoc "" 3 p) in
+      let encoder = encoder "source.dump" Meta_format.empty_metadata in
+      let frame = Frame.create kind in
+      let oc = open_out file in
```

Comment:
Some encoders (currently only `%ffmpeg`) can handle the file outputs themselves. It's important for formats such as `mp4` that need a final header after encoding is finished.

See here: https://github.com/savonet/liquidsoap/blob/master/src/outputs/pipe_output.ml#L384 and here: https://github.com/savonet/liquidsoap/blob/master/src/outputs/pipe_output.ml#L332

Reply:
I am using the same code as `output.file` now. Are you alright with it?